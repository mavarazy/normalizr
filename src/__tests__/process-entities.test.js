// eslint-env jest
import { denormalize, normalize, schema } from '../index';
import {
  generateOrganization,
  generateOrganizationMembership,
  generateUser
} from '@process-street/subgrade/dist/lib/test/generators';

const AuditMetadata = new schema.Object({});
const Organization = new schema.Entity('Organization', { audit: AuditMetadata });
const User = new schema.Entity('User', { audit: AuditMetadata });

AuditMetadata.define({ createdBy: User, updatedBy: User });

const OrganizationMembership = new schema.Entity('OrganizationMembership', {
  organization: Organization,
  user: User
});

describe('Normalizr schema test', () => {
  const userId = 'iMN428lwSWmZRu8ad11Fbg';
  const email = 'test@process.st';
  const organizationId = 'i3FUO85zlSXhFFbZwthHgg';
  const organizationMembershipId = 'gCcQfV06KXig7iIUyqVFLw';

  beforeEach(() => {
    Date.now = jest.fn(() => 0);
  });

  test('should normalize referenced values', () => {
    const user = generateUser(userId, email);
    const organization = generateOrganization(organizationId);
    const organizationMembership = {
      ...generateOrganizationMembership(userId, organizationId),
      id: organizationMembershipId,
      user,
      organization
    };

    const normalized = normalize(organizationMembership, OrganizationMembership);
    expect(normalized).toMatchSnapshot();
  });

  test('should restore referenced value', () => {
    const user = generateUser(userId, email);
    const organization = generateOrganization(organizationId);
    const organizationMembership = {
      ...generateOrganizationMembership(userId, organizationId),
      id: organizationMembershipId,
      user,
      organization
    };

    const { entities } = normalize(organizationMembership, OrganizationMembership);
    const recreated = denormalize(organizationMembershipId, OrganizationMembership, entities);
    expect(recreated).toMatchSnapshot();
  });

  test('should keep reference on all entities match', () => {
    const organizationMembership = {
      ...generateOrganizationMembership(userId, organizationId),
      id: organizationMembershipId
    };

    const { entities } = normalize(organizationMembership, OrganizationMembership);
    const recreated = denormalize(organizationMembershipId, OrganizationMembership, entities);
    expect(recreated).toEqual(organizationMembership);
  });

  test('should normalize audit metadata', () => {
    const createdBy = generateUser();
    const updatedBy = generateUser();

    const user = {
      ...generateUser(),
      audit: {
        createdDate: new Date(0),
        createdBy,
        updatedDate: new Date(0),
        updatedBy
      }
    };

    const { entities } = normalize(user, User);

    const restoredUpdatedBy = denormalize(updatedBy.id, User, entities);
    expect(restoredUpdatedBy).toEqual(updatedBy);

    const restoredCreatedBy = denormalize(createdBy.id, User, entities);
    expect(restoredCreatedBy).toEqual(createdBy);

    const restoredUser = denormalize(user.id, User, entities);
    const expectedUser = {
      ...user,
      audit: {
        ...user.audit,
        createdBy: { id: createdBy.id },
        updatedBy: { id: updatedBy.id }
      }
    };
    expect(restoredUser).toEqual(expectedUser);
  });
});
