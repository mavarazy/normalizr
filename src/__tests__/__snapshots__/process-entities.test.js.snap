// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normalizr schema test should normalize referenced values 1`] = `
Object {
  "entities": Object {
    "Organization": Object {
      "i3FUO85zlSXhFFbZwthHgg": Object {
        "audit": Object {
          "createdBy": Object {
            "id": "is8V0vRjko9RIENELxtOfQ",
          },
          "createdDate": 1970-01-01T00:00:00.000Z,
          "updatedBy": Object {
            "id": "is8V0vRjko9RIENELxtOfQ",
          },
          "updatedDate": 1970-01-01T00:00:00.000Z,
        },
        "cardDefined": true,
        "email": "some@process.st",
        "growSumoPartnerKey": "growTest",
        "id": "i3FUO85zlSXhFFbZwthHgg",
        "industry": "Human Resources",
        "minQuantity": 0,
        "name": "Some Organization",
        "settings": Object {},
        "stripeId": "cus_DULho46NcyD1TO",
        "subscription": Object {
          "cancelAtPeriodEnd": false,
          "currentPeriodEndDate": 1970-01-01T00:00:00.000Z,
          "currentPeriodStartDate": 1970-01-01T00:00:00.000Z,
          "plan": Object {
            "id": "hynvsBS2WQ_q9qZPIANOOA",
          },
          "quantity": 6,
          "status": "Active",
        },
      },
    },
    "OrganizationMembership": Object {
      "gCcQfV06KXig7iIUyqVFLw": Object {
        "admin": false,
        "guest": false,
        "id": "gCcQfV06KXig7iIUyqVFLw",
        "organization": Object {
          "id": "i3FUO85zlSXhFFbZwthHgg",
        },
        "user": Object {
          "id": "iMN428lwSWmZRu8ad11Fbg",
        },
      },
    },
    "User": Object {
      "iMN428lwSWmZRu8ad11Fbg": Object {
        "audit": Object {
          "createdBy": Object {
            "id": "iMN428lwSWmZRu8ad11Fbg",
          },
          "createdDate": 1970-01-01T00:00:00.000Z,
          "updatedBy": Object {
            "id": "iMN428lwSWmZRu8ad11Fbg",
          },
          "updatedDate": 1970-01-01T00:00:00.000Z,
        },
        "beta": false,
        "developer": true,
        "email": "test@process.st",
        "emailVerificationCodeId": "iMN428lwSWmZRu8ad11Fbg",
        "emailVerificationSentDate": 1970-01-01T00:00:00.000Z,
        "emailVerified": true,
        "google": Object {
          "domain": "process.st",
          "email": "test@process.st",
          "id": "101992404834855289963",
        },
        "id": "iMN428lwSWmZRu8ad11Fbg",
        "origin": "FrontStreet",
        "ownership": "User",
        "primaryUserEmail": Object {
          "id": "iMN428lwSWmZRu8ad11Fbg",
        },
        "profile": Object {
          "timeZone": "Canada/Vancouver",
          "username": "Some",
        },
        "signUpProvider": "Google",
        "status": "Active",
        "userType": "Standard",
        "yammer": Object {
          "accessToken": "fake-token",
          "domain": "process.st",
          "email": "test@process.st",
          "id": 1678452238,
        },
      },
    },
  },
  "result": "gCcQfV06KXig7iIUyqVFLw",
}
`;

exports[`Normalizr schema test should restore referenced value 1`] = `
Object {
  "admin": false,
  "guest": false,
  "id": "gCcQfV06KXig7iIUyqVFLw",
  "organization": Object {
    "audit": Object {
      "createdBy": Object {
        "id": "is8V0vRjko9RIENELxtOfQ",
      },
      "createdDate": 1970-01-01T00:00:00.000Z,
      "updatedBy": Object {
        "id": "is8V0vRjko9RIENELxtOfQ",
      },
      "updatedDate": 1970-01-01T00:00:00.000Z,
    },
    "cardDefined": true,
    "email": "some@process.st",
    "growSumoPartnerKey": "growTest",
    "id": "i3FUO85zlSXhFFbZwthHgg",
    "industry": "Human Resources",
    "minQuantity": 0,
    "name": "Some Organization",
    "settings": Object {},
    "stripeId": "cus_DULho46NcyD1TO",
    "subscription": Object {
      "cancelAtPeriodEnd": false,
      "currentPeriodEndDate": 1970-01-01T00:00:00.000Z,
      "currentPeriodStartDate": 1970-01-01T00:00:00.000Z,
      "plan": Object {
        "id": "hynvsBS2WQ_q9qZPIANOOA",
      },
      "quantity": 6,
      "status": "Active",
    },
  },
  "user": Object {
    "audit": Object {
      "createdBy": Object {
        "id": "iMN428lwSWmZRu8ad11Fbg",
      },
      "createdDate": 1970-01-01T00:00:00.000Z,
      "updatedBy": Object {
        "id": "iMN428lwSWmZRu8ad11Fbg",
      },
      "updatedDate": 1970-01-01T00:00:00.000Z,
    },
    "beta": false,
    "developer": true,
    "email": "test@process.st",
    "emailVerificationCodeId": "iMN428lwSWmZRu8ad11Fbg",
    "emailVerificationSentDate": 1970-01-01T00:00:00.000Z,
    "emailVerified": true,
    "google": Object {
      "domain": "process.st",
      "email": "test@process.st",
      "id": "101992404834855289963",
    },
    "id": "iMN428lwSWmZRu8ad11Fbg",
    "origin": "FrontStreet",
    "ownership": "User",
    "primaryUserEmail": Object {
      "id": "iMN428lwSWmZRu8ad11Fbg",
    },
    "profile": Object {
      "timeZone": "Canada/Vancouver",
      "username": "Some",
    },
    "signUpProvider": "Google",
    "status": "Active",
    "userType": "Standard",
    "yammer": Object {
      "accessToken": "fake-token",
      "domain": "process.st",
      "email": "test@process.st",
      "id": 1678452238,
    },
  },
}
`;
